core/txpool/bundlepool/bundlepool.go

```diff
@@ -0,0 +1,442 @@
+package bundlepool
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/prque"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/state"
+	"github.com/ethereum/go-ethereum/core/txpool"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/metrics"
+	"github.com/ethereum/go-ethereum/params"
+	"sync/atomic"
+	"container/heap"
+)
+
+const (
+	// TODO: decide on a good default value
+	// bundleSlotSize is used to calculate how many data slots a single bundle
+	// takes up based on its size. The slots are used as DoS protection, ensuring
+	// that validating a new bundle remains a constant operation (in reality
+	// O(maxslots), where max slots are 4 currently).
+	bundleSlotSize = 128 * 1024 // 128KB
+)
+
+var (
+	bundleGauge = metrics.NewRegisteredGauge("bundlepool/bundles", nil)
+	slotsGauge  = metrics.NewRegisteredGauge("bundlepool/slots", nil)
+)
+
+// BlockChain defines the minimal set of methods needed to back a tx pool with
+// a chain. Exists to allow mocking the live chain out of tests.
+type BlockChain interface {
+	// Config retrieves the chain's fork configuration.
+	Config() *params.ChainConfig
+
+	// CurrentBlock returns the current head of the chain.
+	CurrentBlock() *types.Header
+
+	// GetBlock retrieves a specific block, used during pool resets.
+	GetBlock(hash common.Hash, number uint64) *types.Block
+
+	// StateAt returns a state database for a given root hash (generally the head).
+	StateAt(root common.Hash) (*state.StateDB, error)
+}
+
+type BundleSimulator interface {
+	SimulateBundle(bundle *types.Bundle) (*big.Int, error)
+}
+
+type BundlePool struct {
+	config Config
+	gasTip atomic.Pointer[big.Int] // Currently accepted minimum gas tip
+
+	bundles    map[common.Hash]*types.Bundle
+	bundleHeap BundleHeap
+	mu         sync.RWMutex
+
+	slots uint64 // Number of slots currently allocated
+
+	bundleGasPricer *BundleGasPricer
```

Comment:
emm, can we reuse the bundle pool in the `DirectRoute`..

Reply:
You mean the gas part? Yes I reuse most of it